include/aws/crt/io/Stream.h

```diff
@@ -16,21 +16,114 @@
 
 #include <aws/crt/Exports.h>
 #include <aws/crt/Types.h>
-
-struct aws_input_stream;
+#include <aws/io/stream.h>
 
 namespace Aws
 {
     namespace Crt
     {
         namespace Io
         {
-            /**
-             * Factory to create a aws-c-io input stream subclass from a C++ stream
+            using StreamStatus = aws_stream_status;
+            using OffsetType = aws_off_t;
+
+            enum class StreamSeekBasis
+            {
+                Begin = AWS_SSB_BEGIN,
+                End = AWS_SSB_END,
+            };
+
+            /***
+             * Interface for building an Object oriented stream that will be honored by the CRT's low-level
+             * aws_input_stream interface. To use, create a subclass of InputStream and define the abstract
+             * functions.
              */
-            AWS_CRT_CPP_API aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
-                Aws::Crt::Allocator *allocator = DefaultAllocator()) noexcept;
+            class AWS_CRT_CPP_API InputStream
+            {
+              public:
+                virtual ~InputStream();
+
+                InputStream(const InputStream &) = delete;
+                InputStream &operator=(const InputStream &) = delete;
+                InputStream(InputStream &&) = delete;
+                InputStream &operator=(InputStream &&) = delete;
+
+                explicit operator bool() const noexcept { return IsGood(); }
+
+                bool IsGood() const noexcept;
+
+                aws_input_stream *GetUnderlyingStream() noexcept { return &m_underlying_stream; }
+
+              protected:
+                Allocator *m_allocator;
+                aws_input_stream m_underlying_stream;
+                bool m_good;
+
+                InputStream(Aws::Crt::Allocator *allocator = DefaultAllocator());
+
+                /***
+                 * Read up-to buffer::capacity - buffer::len into buffer::buffer
+                 * Increment buffer::len by the amount you read in.
+                 *
+                 * @return true on success, false otherwise. Return false, when there is nothing left to read.
+                 * You SHOULD raise an error via aws_raise_error()
+                 * if an actual failure condition occurs.
+                 * If you return false, GetStatusImpl() will be called to determine
+                 * the validity of the stream.
+                 */
+                virtual bool ReadImpl(ByteBuf &buffer) noexcept = 0;
```

Comment:
why do we do check GetStatusImpl() when ReadImpl() fails, but we just set straight up set m_good=false when SeekImpl() fails?